Guesswho book version

matthew-goh opened the pull request.

matthew-goh committed:
Merge pull request #3 from matthew-goh/main

Reverting changes temporarily

matthew-goh committed:
All book version updates for PR

simondrugan16 commented on src/main/scala/guessbook/BookGuessChecker.scala:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  implicit object CountryGuessChecker extends BookGuessChecker[Country.Value] {
+    def validateGuess(guessAttribute: BookAttribute, guessValue: Country.Value) : Unit = {
```
is guessValue being used?

matthew-goh replied:
Indeed, guessValue actually isn't needed here since its type is already used by poseQuestion() to determine which implicit object to use

simondrugan16 commented on src/main/scala/guessbook/BookGuessChecker.scala:
```diff
@@ -0,0 +1,99 @@
+      }
+    }
+
+    def printFeedback(selectedBook: Book, guessAttribute: BookAttribute, guessValue: SuspectPool.Value): Unit = {
```
could these printFeedback be extracted to reduce repeating the code multiple times?

matthew-goh replied:
Moved the default implementation of printFeedback() into the trait (and the implicit objects override it if needed)

simondrugan16 commented on src/main/scala/guessbook/GuessBook.scala:
```diff
@@ -0,0 +1,152 @@
+
+object GuessBook {
+  // instantiate all books
+  val TheHonjinMurders: Book = Book(title = "The Honjin Murders", attributes = Map(CountryOfOrigin -> Country.Japan, Honkaku -> true, AuthorGender -> GenderEnum.Male,
```
please move possible data items to a separate file

simondrugan16 commented on src/main/scala/guessbook/GuessBook.scala:
```diff
@@ -0,0 +1,152 @@
+object GuessBook {
+  // instantiate all books
+  val TheHonjinMurders: Book = Book(title = "The Honjin Murders", attributes = Map(CountryOfOrigin -> Country.Japan, Honkaku -> true, AuthorGender -> GenderEnum.Male,
+    SchoolOrUniversityStudents -> false, DetectiveType -> Detective.PrivateProfessional, Series -> true, PushkinVertigo -> true, SuspectPoolType -> SuspectPool.RestrictedCommunity))
```
an extension task (from Aashvin) would be to use a csv (or any other) file that would hold all this data, and could be parsed dynamically into books at the start of the programme

matthew-goh replied:
Created method getBooksFromCSV() to do this

simondrugan16 commented on src/main/scala/guessbook/BookGuessChecker.scala:
```diff
@@ -0,0 +1,99 @@
+  implicit object GenderGuessChecker extends BookGuessChecker[GenderEnum.Value] {
+    def validateGuess(guessAttribute: BookAttribute, guessValue: GenderEnum.Value) : Unit = {
+      guessAttribute match {
+        case AuthorGender =>
```
please explicitly return unit here rather than blank

simondrugan16 replied:
for clarity

simondrugan16 commented on src/main/scala/guessbook/GuessBook.scala:
```diff
@@ -0,0 +1,152 @@
+                         remainingBooks: Seq[Book]): Seq[Book] = {
+      if (selectedBook.attributes(guessAttribute) == guessValue){
+        remainingBooks.filter {
+          ch => ch.attributes(guessAttribute) == guessValue
```
please rename ch to what it is

simondrugan16 commented on src/main/scala/guessbook/GuessBook.scala:
```diff
@@ -0,0 +1,152 @@
+        remainingBooks.filter {
+          ch => ch.attributes(guessAttribute) == guessValue
+        }
+      } else { // keep characters without the guessed value
```
i like the logic here

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+    val game = GuessBook
+    println("START OF GAME\n")
+
+    // Randomly select a book for each player
```
could you ask the user which mode they want to play it on easily (either random book, or player vs player)?

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+    var gameOver = false
+    var guesser = 2
+
+    while (!gameOver) {
```
is it possible to replace this with recursion, and also remove the vars by passing them through with recursion (and instantiate them initially by using default values in the method argument?

matthew-goh replied:
Yes, now each player turn is a recursive call and guesser determines which board is queried and updated. gameOver is not needed anymore since the method terminates (no more recursive call) if that case is reached

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+              val guessValue: Country.Value = specifyEnumValue(Country)
+              booksOnBoard(guesser-1) = game.poseQuestion(selectedBooks(guesser-1), guessAttribute, guessValue, booksOnBoard(guesser-1))
+            }
+            case DetectiveType => {
```
maybe `guesser-1` can be extracted somewhere so that it can be represented as like _otherPlayer_?

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+        case 2 => { // guessing book title
+          val guessedTitle: String = specifyBookTitle()
+          val guessResult = game.guessBookTitle(selectedBooks(guesser-1), guessedTitle, guesser, booksOnBoard(guesser-1))
+          booksOnBoard(guesser-1) = guessResult._1
```
is this mutability?

matthew-goh replied:
Yes, booksOnBoard is an Array, which is mutable even though it's declared as a val. So that regardless of which player is the guesser, booksOnBoard can be updated using the same line of code and passed into the next recursive call (if needed)

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+          gameOver = guessResult._2
+          if (!gameOver){
+            println(s"Books remaining on P$guesser's board:")
+            game.printRemainingBooks(booksOnBoard(guesser-1), maxBooksPerLine = 5)
```
is maxBooksPerLine used to emulate a guess who (or guess what in this case :D ) board?

matthew-goh replied:
It does have that effect! But I added it so that the list of books wouldn't go off the screen

simondrugan16 replied:
In that case, would it make sense to limit it by total characters rather than number of books?

matthew-goh replied:
Yes, that would make the lines more even... although using number of books could help the player to easily count how many books are left on the board. Still, I've created a new version of the function with maxCharsPerLine

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+    game.printRemainingBooks(game.allBooks, maxBooksPerLine = 5)
+    println()
+
+    while (matchingBook.isEmpty) {
```
again can this be replaced with recursion?

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+
+  def chooseAttributeOrTitle(): Int = {
+    var input: String = ""
+    var validInput = false
```
VARS?!?!?

matthew-goh replied:
Changed all such methods taking user input to use recursion!

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+    guessInput.toLowerCase.toBoolean
+  }
+
+  def specifyEnumValue[E <: Enumeration with BookAttributeEnum](attributeEnum: E): attributeEnum.Value = {
```
love enums here

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+    var validInput = false
+
+    // println(attributeEnum.values.zipWithIndex)
+    val selectOptionStr: String = attributeEnum.values.zipWithIndex.map {
```
love zipWithIndex use

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,206 @@
+      }
+    }
+
+    val selectedIndex = input.toInt - 1
```
inline this in below line?

simondrugan16 commented on src/main/scala/guesswho/GameCharacter.scala:
```diff
@@ -1,3 +1,3 @@
 package guesswho
 
-case class GameCharacter(val name: String, val attributes: Map[Attribute, Any])
+case class GameCharacter(name: String, attributes: Map[Attribute, Any])
```
what's the difference between having vals here or not

matthew-goh replied:
No difference to the functionality, but case class doesn't need val and I forgot to remove it when I changed it from a class to case class

simondrugan16 replied:
When injecting with Play in a class you can use/not use the val X: Y too. Do you know what the difference in functionality here would be? (it seems to not make a difference for case classes)

matthew-goh replied:
For a normal class, not putting val means the parameter is a constructor parameter but not a field, so it can't be accessed from outside the class

simondrugan16 commented on src/test/scala/guessbook/GuessBookSpec.scala:
```diff
@@ -0,0 +1,130 @@
+      assert(game.DanganronpaKirigiri.attributes(SuspectPoolType) === SuspectPool.ClosedCircle)
+    }
+
+    "create a sorted list of all book titles" in {
```
is there a reason for them to be sorted?

matthew-goh replied:
I feel that when printing the remaining books, it's better to ensure that they appear in a clear and consistent order

simondrugan16 commented on src/test/scala/guessbook/GuessBookSpec.scala:
```diff
@@ -0,0 +1,130 @@
+  "printRemainingBooks()" should {
+    "print the list of book titles, with the correct number of titles on each line" in {
+      val stream = new java.io.ByteArrayOutputStream()
+      Console.withOut(stream) {
```
i like the effort put into testing these console in/outs :D

simondrugan16 commented on src/test/scala/guessbook/GuessBookSpec.scala:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    "throw an exception if the attribute is invalid for the guess value type" in {
+      val invalidGuess1 = intercept[Exception](game.poseQuestion(game.Malice, CountryOfOrigin, true, game.allBooks))
```
i like the use of intercept here

simondrugan16 commented on src/test/scala/guessbook/GuessBookSpec.scala:
```diff
@@ -0,0 +1,130 @@
+      assert(stream.toString.strip === "Player 2's guess is incorrect. The book is not Death Among the Undead.")
+    }
+
+    "throw an exception if an invalid title is input" in {
```
why throw an exception here?

matthew-goh replied:
Since this method isn't the one that accepts and validates user input, but the one that filters based on the guess, the game shouldn't proceed if an invalid value has reached it (similar to the exceptions thrown for invalid guess attribute in poseQuestion)

simondrugan16 commented on src/test/scala/guessbook/GuessBookUserInputSpec.scala:
```diff
@@ -0,0 +1,224 @@
+          assert(GuessBookUserInput.specifyEnumValue(Detective) == Detective.OrdinaryPerson)
+        }
+      }
+      outputStream.toString should include ("Enter [1] for police\n" +
```
why _include_ not _be_

matthew-goh replied:
The part of the console output being checked might not be everything printed, e.g. specifyEnumValue() also prints "Select a value to guess:"

matthew-goh committed:
Some edits in response to PR review

- Move instantiated book info to another file
- Default printFeedback() implementation in trait
- Can choose whether selected books are random
- Use recursion in methods asking for user input

matthew-goh committed:
Use recursion for player turns

matthew-goh committed:
Read book info from csv file

matthew-goh committed:
Put reading from csv in a method

+ Only instantiate GuessBook once in GuessBookUserInput

matthew-goh committed:
guessValue not needed in validateGuess; use for comprehension in getBooksFromCSV

simondrugan16 commented on src/main/scala/guessbook/GuessBook.scala:
```diff
@@ -0,0 +1,111 @@
+
+import scala.util.{Failure, Random, Success, Try}
+
+object GuessBook {
```
Do you think it would make sense to make this into a class? Then the getBooksFromCSV (anything to do with parsing files/reading data) can be moved to a different class and then injected in via DI. This would separate out these things which have different concerns

simondrugan16 commented on src/main/scala/guessbook/GuessBook.scala:
```diff
@@ -0,0 +1,111 @@
+object GuessBook {
+
+  // Read book info from csv file
+  def getBooksFromCSV(path: String): Seq[Book] = {
```
I've never done anything like this so can't say for certain, but it looks good to me, especially the for comp in here :D

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,225 @@
+    getBookInGame(nonGuesser)
+  }
+
+  @tailrec
+  def chooseGameMode(): Int = {
+    println("Enter [1] for each player's selected book to be drawn randomly\n" +
+      "Enter [2] for each player to choose the book for their opponent to guess")
+
+    val input = StdIn.readLine()
+    Try(input.toInt) match {
+      case Success(x) if (x == 1 || x == 2) => x
+      case _ => {
+        println("Invalid input. Please try again.\n")
+        chooseGameMode()
+      }
+    }
+  }
```
Nicely done

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,225 @@
+    }
+  }
+
+  @tailrec
```
I'm loving all this tail recursion

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,225 @@
+    println("Which do you think is the selected book?")
+    val titleInput = StdIn.readLine()
+    val matchingTitle: Option[String] = game.allTitles.find(title => title.toLowerCase == titleInput.toLowerCase)
+    matchingTitle match {
```
if you inline matchingTitle here then this could be tail recursive. I think it's ok to inline this since matchingTitle as a val isn't tooooooo complex, but ofc it's a trade off, so don't do this if you think it's better to have matchingTitle separate

matthew-goh replied:
Actually specifyBookTitle() is already tail recursive since the recursive call is just the function itself - I just hadn't added the @tailrec annotation!

simondrugan16 commented on src/main/scala/guessbook/GuessBookUserInput.scala:
```diff
@@ -0,0 +1,225 @@
+    // play the game
+    playerTurns(selectedBooks, booksOnBoard)
+
+    // recursive function that calls itself until a player guesses correctly
+    @tailrec
+    def playerTurns(selectedBooks: List[Book], booksOnBoard: Array[Seq[Book]], guesser: Int = 1) : Unit = {
```
brilliant

matthew-goh committed:
Edits for 2nd round of PR comments

- getBooksFromCSV in a separate class and injected into GuessBook class
- printRemainingBooks uses maxCharsPerLine

matthew-goh committed:
Add Try for opening csv file from path

+ Strip whitespace from user inputs

matthew-goh committed:
Use latest version of scala-csv
